README.md
#  inline_dialogs

A lightweight Flutter package to implement inline dialogs which can be managed from the main context.
The dialog can be closed from the same code from which it was invoked. 
It also gives back return values on closure helping the developer catch user action on the dialog.

## Getting Started

1. Add below builder code in MaterialApp() in main.dart page [sample](https://github.com/1SouravGhosh/flutter_inline_dialogs/blob/master/example/sample_app.dart/)

<pre>
<code class='language-dart hljs'>
 builder: (context, widget) => Navigator(
        onGenerateRoute: (settings) => MaterialPageRoute(
            builder: (context) => DialogManager(
                  child: widget,
                )),
      ),

</code>
</pre>

2. Add dialogSetupLocator in main() in main.dart page [sample](https://github.com/1SouravGhosh/flutter_inline_dialogs/blob/master/example/sample_app.dart/)

<pre>
<code class='language-dart hljs'>
 void main() {
  dialogSetupLocator();
  runApp(MyApp());
}

</code>
</pre>

That's it! You are good to go. 

## Demo
<p><img src='https://raw.githubusercontent.com/1SouravGhosh/flutter_inline_dialogs/master/assets/explainer.gif' style='width: 30vw; min-width: 30px;' alt='Inline dialog demo'/>
</p>

# Implementation


1. Implement option dialog  [ dialogType: DialogType.option ]
<pre>
<code class='language-dart hljs'>
 var _dialogResponse = await _dialogService.showDialog(
                      title: Icon(
                        Icons.code,
                        size: 50,
                        color: Colors.orange,
                      ),
                      content: Padding(
                        padding: const EdgeInsets.all(12.0),
                        child: Text(
                          'Sample option dialog',
                          style: Theme.of(context).textTheme.headline6,
                        ),
                      ),
                      dialogType: DialogType.option,
                      optionRight: 'Right',
                      optionLeft: 'Left');
                 
                    
print(_dialogResponse.optionRight);
print(_dialogResponse.optionLeft);
                  
</code>
</pre>

2. Implement confirm dialog  [ dialogType: DialogType.confirm ]
<pre>
<code class='language-dart hljs'>
 var _dialogResponse = await _dialogService.showDialog(
                      title: Icon(
                        Icons.done_outline,
                        size: 50,
                        color: Colors.green,
                      ),
                      content: Padding(
                        padding: const EdgeInsets.all(12.0),
                        child: Text(
                          'Sample confirm dialog',
                          style: Theme.of(context).textTheme.headline6,
                        ),
                      ),
                      dialogType: DialogType.confirm,
                      buttonText: 'Done');
                 
                    setState(() {
                      _confirmBtn =  _dialogResponse.confirmed;
                  });
                                  
print('${_dialogResponse.confirmed}');
                  
</code>
</pre>

3. Implement waiter dialog  [ dialogType: DialogType.waiter]
<pre>
<code class='language-dart hljs'>
 _dialogService.showDialog(
                      title: Text(
                        'Sample waiter dialog',
                        style: Theme.of(context).textTheme.headline6,
                      ),
                      content: Padding(
                        padding: const EdgeInsets.all(12.0),
                        child: CupertinoActivityIndicator(
                          radius: 20,
                        ),
                      ),
                      dialogType: DialogType.waiter);
                  await Future.delayed(Duration(seconds: 5));
                  _dialogService.dismissDialog();               
</code>
</pre>



You can close all of the above dialogs by using   _dialogService.dismissDialog(); from the same context where the dialog was invoked.

## Credit

A special shout out to FilledStacks YouTube channel. Please show some love and subscribe to this channel.
[YouTube Link](https://www.youtube.com/watch?v=IrFU_BrCWnE)

**The code has been modified heavily for ease of modular implementation and more features have been added.

## Documentation
For help getting started with Flutter, view our 
[github repo](https://github.com/1SouravGhosh/flutter_inline_dialogs/), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.





CHANGELOG.md
## [1.0.1] - 20th July 2020

* This is the initial release. 
* The package is stable and same code being used used in multiple production app. 

## [1.0.2] - 21th July 2020

* Upgraded dependencies. 
* Added example app. 


lib/dialogs/manager.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:inline_dialogs/custom_utils/service_locator.dart';
import 'package:inline_dialogs/dialogs/model.dart';
import 'package:inline_dialogs/dialogs/service.dart';

class DialogManager extends StatefulWidget {
  final Widget child;
  DialogManager({Key key, this.child}) : super(key: key);

  _DialogManagerState createState() => _DialogManagerState();
}

class _DialogManagerState extends State<DialogManager> {
  DialogService _dialogService = locator<DialogService>();

  @override
  void initState() {
    super.initState();
    _dialogService.registerDialogListener(_showDialog, _dismissDialog);
  }

  @override
  Widget build(BuildContext context) {
    return widget.child;
  }

  void _dismissDialog(DialogRequest request) {
    Navigator.of(context).pop();
  }

  void _showDialog(DialogRequest request) {
    showDialog(
        context: context,
        barrierDismissible: request.dialogType != DialogType.waiter,
        builder: (context) => CupertinoAlertDialog(
              title: request.title,
              content: request.description,
              actions: _buildButton(
                  dialogType: request.dialogType,
                  optionLeft: request.optionLeft,
                  optionRight: request.optionRight,
                  buttonText: request.buttonText),
            ));
  }

  List<CupertinoDialogAction> _buildButton(
      {DialogType dialogType,
      String optionLeft,
      String optionRight,
      String buttonText}) {
    switch (dialogType) {
      case DialogType.option:
        {
          return [
            CupertinoDialogAction(
              child: Text(
                optionLeft,
                style: TextStyle(color: Colors.blueAccent),
              ),
              onPressed: () {
                _dialogService.dialogComplete(
                    DialogResponse(optionLeft: optionLeft, confirmed: true));
                Navigator.of(context).pop();
              },
            ),
            CupertinoDialogAction(
              child: Text(
                optionRight,
                style: TextStyle(color: Colors.red),
              ),
              onPressed: () {
                _dialogService.dialogComplete(
                    DialogResponse(optionRight: optionRight, confirmed: true));
                Navigator.of(context).pop();
              },
            )
          ];
        }
        break;
      case DialogType.confirm:
        {
          return [
            CupertinoDialogAction(
              child: Padding(
                padding: const EdgeInsets.all(10.0),
                child: Text(
                  buttonText,
                  style: TextStyle(color: Colors.blueAccent),
                ),
              ),
              onPressed: () {
                _dialogService.dialogComplete(
                    DialogResponse(optionLeft: buttonText, confirmed: true));
                Navigator.of(context).pop();
              },
            )
          ];
        }
        break;
      case DialogType.waiter:
        {
          return [];
        }
        break;
    }
    return [];
  }
}


lib/dialogs/service.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:inline_dialogs/dialogs/model.dart';

class DialogService {
  Function(DialogRequest) _showDialogListener;
  Function(DialogRequest) _dismissDialogListener;
  Completer<DialogResponse> _dialogCompleter;

  void registerDialogListener(
    Function(DialogRequest) showDialogListener,
    Function(DialogRequest) dismissDialogListener,
  ) {
    _showDialogListener = showDialogListener;
    _dismissDialogListener = dismissDialogListener;
  }

  Future<DialogResponse> showDialog(
      {@required Widget title,
      @required Widget content,
      DialogType dialogType = DialogType.waiter,
      String optionLeft = "Update",
      String optionRight = "Delete",
      String buttonText = "Okay"}) {
    _dialogCompleter = Completer<DialogResponse>();
    _showDialogListener(DialogRequest(
        title: title,
        description: content,
        buttonText: buttonText,
        dialogType: dialogType,
        optionLeft: optionLeft,
        optionRight: optionRight));
    return _dialogCompleter.future;
  }

  dismissDialog() {
    _dismissDialogListener(DialogRequest());
  }

  void dialogComplete(DialogResponse response) {
    _dialogCompleter.complete(response);
  }
}


example/sample_app.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:inline_dialogs/custom_utils/service_locator.dart';
import 'package:inline_dialogs/dialogs/manager.dart';
import 'package:inline_dialogs/dialogs/model.dart';
import 'package:inline_dialogs/dialogs/service.dart';

void main() {
  dialogSetupLocator();
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Inline Dialogs Demo',
      debugShowCheckedModeBanner: false,
      builder: (context, widget) => Navigator(
        onGenerateRoute: (settings) => MaterialPageRoute(
            builder: (context) => DialogManager(
                  child: widget,
                )),
      ),
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Inline Dialogs Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _rightBtn, _leftBtn = "";
  bool _showOptionText = false, _showConfirmText = false, _confirmBtn;
  DialogService _dialogService = locator<DialogService>();
  Timer _counter;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: RaisedButton(
                child: Text(" Option Dialog "),
                onPressed: () {
                  setState(() {
                    _showOptionText = true;
                    _showConfirmText = false;
                    _rightBtn = null;
                    _leftBtn = null;
                  });
                  var _dialogResponse = _dialogService.showDialog(
                      title: Icon(
                        Icons.code,
                        size: 50,
                        color: Colors.orange,
                      ),
                      content: Padding(
                        padding: const EdgeInsets.all(12.0),
                        child: Text(
                          "Sample option dialog",
                          style: Theme.of(context).textTheme.headline6,
                        ),
                      ),
                      dialogType: DialogType.option,
                      optionRight: "Right",
                      optionLeft: "Left");
                  _dialogResponse.then((value) {
                    setState(() {
                      _rightBtn = value.optionRight;
                      _leftBtn = value.optionLeft;
                    });
                  });
                },
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: RaisedButton(
                child: Text("Confirm Dialog"),
                onPressed: () {
                  setState(() {
                    _showConfirmText = true;
                    _showOptionText = false;
                    _confirmBtn = false;
                  });
                  var _dialogResponse = _dialogService.showDialog(
                      title: Icon(
                        Icons.done_outline,
                        size: 50,
                        color: Colors.green,
                      ),
                      content: Padding(
                        padding: const EdgeInsets.all(12.0),
                        child: Text(
                          "Sample confirm dialog",
                          style: Theme.of(context).textTheme.headline6,
                        ),
                      ),
                      dialogType: DialogType.confirm,
                      buttonText: "Done");
                  _dialogResponse.then((value) {
                    setState(() {
                      _confirmBtn = value.confirmed;
                    });
                  });
                },
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: RaisedButton(
                child: Text(" Waiter Dialog  "),
                onPressed: () async {
                  setState(() {
                    _showConfirmText = false;
                    _showOptionText = false;
                  });
                  var _timer = Timer.periodic(Duration(seconds: 1), (timer) {
                    setState(() {
                      _counter = timer;
                    });
                  });
                  _dialogService.showDialog(
                    title: Text(
                      "Sample waiter dialog",
                      style: Theme.of(context).textTheme.headline6,
                    ),
                    content: Padding(
                      padding: const EdgeInsets.all(12.0),
                      child: CupertinoActivityIndicator(
                        radius: 20,
                      ),
                    ),
                  );
                  await Future.delayed(Duration(seconds: 5));
                  setState(() {
                    _dialogService.dismissDialog();
                  });
                  _timer.cancel();
                },
              ),
            ),
            _showOptionText
                ? Padding(
                    padding: const EdgeInsets.only(top: 160.0),
                    child: Text(
                      'Left btn returns : $_leftBtn',
                      style: Theme.of(context).textTheme.headline6,
                      textAlign: TextAlign.center,
                    ),
                  )
                : Container(),
            _showOptionText
                ? Padding(
                    padding: const EdgeInsets.only(top: 10.0),
                    child: Text(
                      'Right btn Returns : $_rightBtn',
                      style: Theme.of(context).textTheme.headline6,
                      textAlign: TextAlign.center,
                    ),
                  )
                : Container(),
            _showConfirmText
                ? Padding(
                    padding: const EdgeInsets.only(top: 160.0),
                    child: Text(
                      'Confirm btn returns : $_confirmBtn',
                      style: Theme.of(context).textTheme.headline6,
                      textAlign: TextAlign.center,
                    ),
                  )
                : Container(),
            !_showConfirmText && !_showOptionText
                ? Padding(
                    padding: const EdgeInsets.only(top: 180.0),
                    child: Text(
                      'Counter: ${_counter?.tick}',
                      style: Theme.of(context).textTheme.headline4,
                      textAlign: TextAlign.center,
                    ),
                  )
                : Container(),
          ],
        ),
      ),
      // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
